module/zfs/zio.c

```diff
@@ -3110,10 +3137,10 @@ zio_write_gang_member_ready(zio_t *zio)
 	VERIFY3U(BP_GET_NDVAS(zio->io_bp), <=, BP_GET_NDVAS(pio->io_bp));
 
 	mutex_enter(&pio->io_lock);
-	for (int d = 0; d < BP_GET_NDVAS(zio->io_bp); d++) {
+	for (int d = 0; d < BP_GET_NDVAS(pio->io_bp); d++) {
 		ASSERT(DVA_GET_GANG(&pdva[d]));
 		asize = DVA_GET_ASIZE(&pdva[d]);
-		asize += DVA_GET_ASIZE(&cdva[d]);
+		asize += DVA_GET_ASIZE(&cdva[0]);
 		DVA_SET_ASIZE(&pdva[d], asize);
 	}
```

Comment:
I haven't looked close how ASIZE supposed to be accounted gangs and may be wrong, but I have feeling that neither old nor new code here is correct when number of copies for the parent and the child blocks is different.  The old code obviously could get out of array, while the new seems to account only `ASIZE` of the first `DVA`, but not others.

Reply:
Well, the VERIFY just before this loop (which checks exactly that) hasn't tripped in any of my testing, so I suspect the assumption still holds. I can try to come up with some cases where it wouldn't, though.

Reply:
I think the only way it can happen right now is with some abuse of the `dmu_ddt_copies` tunable; if you set the redundant metadata level low and then force that tunable to 3, then you might have more leaf copies than gang copies. But that wouldn't be a regression in this code, it would hit the VERIFY on master too. We could add a line in `dmu_write_policy` to set gang_copies to always be at least as high as `copies`?

Reply:
We could do that, I think that would work. The problem with this in general is that if you set the `dmu_ddt_copies` property high, you presumably really want that data to be reliable. And allowing the gang copies to be lower than the normal copies is not giving the user the reliability they asked for. So I think independently of this change, we should probably ensure that gang_copies is always at least as high as copies (in `dmu_write_policy`; it can be 0 in the DDT case, but in that case we are forbidden from ganging, and this case isn't relevant).